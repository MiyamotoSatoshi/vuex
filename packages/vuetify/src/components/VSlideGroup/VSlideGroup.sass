// Imports
@import '../../styles/styles.sass'
@import './_variables.scss'

// Block
.v-slide-group
  display: flex

.v-slide-item
  display: inline-flex
  flex: 0 1 auto

// Element
.v-slide-group__append,
.v-slide-group__prepend
  align-items: center
  display: flex
  flex: 0 1 $slide-group-prepend-basis
  justify-content: center
  min-width: $slide-group-prepend-basis

.v-slide-group__content
  display: flex
  flex: 1 0 auto
  position: relative
  transition: $primary-transition
  white-space: nowrap

.v-slide-group__wrapper
  contain: content
  display: flex
  overflow: hidden

// Modifiers
.v-slide-group__append,
.v-slide-group__prepend
  &--disabled
    pointer-events: none
