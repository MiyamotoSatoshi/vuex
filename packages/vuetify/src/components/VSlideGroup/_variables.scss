$slide-group-prepend-basis: 48px !default;
