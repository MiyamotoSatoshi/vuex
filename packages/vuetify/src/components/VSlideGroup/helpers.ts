export function bias (val: number) {
  const c = 0.501
  const x = Math.abs(val)
  return Math.sign(val) * (x / ((1 / c - 2) * (1 - x) + 1))
}

export function calculateUpdatedOffset ({
  selectedElement,
  containerSize,
  contentSize,
  isRtl,
  currentScrollOffset,
  isHorizontal,
}: {
  selectedElement: HTMLElement
  containerSize: number
  contentSize: number
  isRtl: boolean
  currentScrollOffset: number
  isHorizontal: boolean
}): number {
  const clientSize = isHorizontal ? selectedElement.clientWidth : selectedElement.clientHeight
  const offsetStart = isHorizontal ? selectedElement.offsetLeft : selectedElement.offsetTop
  const adjustedOffsetStart = isRtl && isHorizontal ? (contentSize - offsetStart - clientSize) : offsetStart

  const totalSize = containerSize + currentScrollOffset
  const itemOffset = clientSize + adjustedOffsetStart
  const additionalOffset = clientSize * 0.4

  if (adjustedOffsetStart <= currentScrollOffset) {
    currentScrollOffset = Math.max(adjustedOffsetStart - additionalOffset, 0)
  } else if (totalSize <= itemOffset) {
    currentScrollOffset = Math.min(currentScrollOffset - (totalSize - itemOffset - additionalOffset), contentSize - containerSize)
  }

  return currentScrollOffset
}

export function calculateCenteredOffset ({
  selectedElement,
  containerSize,
  contentSize,
  isRtl,
  isHorizontal,
}: {
  selectedElement: HTMLElement
  containerSize: number
  contentSize: number
  isRtl: boolean
  isHorizontal: boolean
}): number {
  const clientSize = isHorizontal ? selectedElement.clientWidth : selectedElement.clientHeight
  const offsetStart = isHorizontal ? selectedElement.offsetLeft : selectedElement.offsetTop

  const offsetCentered = isRtl && isHorizontal
    ? contentSize - offsetStart - clientSize / 2 - containerSize / 2
    : offsetStart + clientSize / 2 - containerSize / 2

  return Math.min(contentSize - containerSize, Math.max(0, offsetCentered))
}
