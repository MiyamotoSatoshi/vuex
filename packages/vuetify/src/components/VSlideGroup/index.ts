export { VSlideGroup } from './VSlideGroup'
export { VSlideGroupItem } from './VSlideGroupItem'
