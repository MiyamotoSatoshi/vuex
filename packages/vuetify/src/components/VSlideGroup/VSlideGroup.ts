// Styles
import './VSlideGroup.sass'

// Components
import VIcon from '../VIcon'
import { VFadeTransition } from '../transitions'

// Extensions
import { BaseItemGroup } from '../VItemGroup/VItemGroup'

// Directives
import Resize from '../../directives/resize'
import Touch from '../../directives/touch'

// Utilities
import mixins, { ExtractVue } from '../../util/mixins'

// Types
import Vue, { VNode } from 'vue'

interface TouchEvent {
  touchstartX: number
  touchmoveX: number
}

interface options extends Vue {
  $refs: {
    content: HTMLElement
    wrapper: HTMLElement
  }
}

export const BaseSlideGroup = mixins<options &
/* eslint-disable indent */
  ExtractVue<typeof BaseItemGroup>
/* eslint-enable indent */
>(
  BaseItemGroup
  /* @vue/component */
).extend({
  name: 'base-slide-group',

  directives: {
    Resize,
    Touch
  },

  props: {
    activeClass: {
      type: String,
      default: 'v-slide-item--active'
    },
    appendIcon: {
      type: String,
      default: '$vuetify.icons.next'
    },
    mobileBreakPoint: {
      type: [Number, String],
      default: 1264,
      validator: (v: any) => !isNaN(parseInt(v))
    },
    prependIcon: {
      type: String,
      default: '$vuetify.icons.prev'
    },
    showArrows: Boolean
  },

  data: () => ({
    isOverflowing: false,
    resizeTimeout: 0,
    startX: 0,
    scrollOffset: 0,
    widths: {
      content: 0,
      wrapper: 0
    }
  }),

  computed: {
    __cachedAppend (): VNode {
      return this.genTransition('append')
    },
    __cachedPrepend (): VNode {
      return this.genTransition('prepend')
    },
    classes (): object {
      return BaseItemGroup.options.computed.classes.call(this)
    },
    hasAffixes (): Boolean {
      return (
        (this.showArrows || !this.isMobile) &&
        this.isOverflowing
      )
    },
    hasAppend (): boolean {
      if (!this.hasAffixes) return false

      const { content, wrapper } = this.widths

      // Check one scroll ahead to know the width of right-most item
      return content > this.scrollOffset + wrapper
    },
    hasPrepend (): boolean {
      return this.hasAffixes && this.scrollOffset > 0
    },
    isMobile (): boolean {
      return this.$vuetify.breakpoint.width < this.mobileBreakPoint
    }
  },

  watch: {
    scrollOffset (val) {
      this.$refs.content.style.transform = `translateX(${-val}px)`
    },
    widths: 'setOverflow'
  },

  methods: {
    genAppend (): VNode | null {
      if (!this.isOverflowing && !this.showArrows) return null

      const slot = this.$scopedSlots.append
        ? this.$scopedSlots.append({})
        : this.$slots.append || this.__cachedAppend

      return this.$createElement('div', {
        staticClass: 'v-slide-group__append',
        class: {
          'v-slide-group__append--disabled': !this.hasAppend
        },
        on: {
          click: () => this.onAffixClick('append')
        },
        key: 'append'
      }, [slot])
    },
    genContent (): VNode {
      return this.$createElement('div', {
        staticClass: 'v-slide-group__content',
        ref: 'content'
      }, this.$slots.default)
    },
    genIcon (location: 'prepend' | 'append'): VNode | null {
      const upperLocation = `${location[0].toUpperCase()}${location.slice(1)}`
      const hasAffix = (this as any)[`has${upperLocation}`]

      if (
        !this.showArrows &&
        !hasAffix
      ) return null

      return this.$createElement(VIcon, {
        props: {
          disabled: !hasAffix
        }
      }, (this as any)[`${location}Icon`])
    },
    genPrepend (): VNode | null {
      if (!this.isOverflowing && !this.showArrows) return null

      const slot = this.$scopedSlots.prepend
        ? this.$scopedSlots.prepend({})
        : this.$slots.prepend || this.__cachedPrepend

      return this.$createElement('div', {
        staticClass: 'v-slide-group__prepend',
        class: {
          'v-slide-group__prepend--disabled': !this.hasPrepend
        },
        on: {
          click: () => this.onAffixClick('prepend')
        },
        key: 'prepend'
      }, [slot])
    },
    genTransition (location: 'prepend' | 'append') {
      return this.$createElement(VFadeTransition, [this.genIcon(location)])
    },
    genWrapper (): VNode {
      return this.$createElement('div', {
        staticClass: 'v-slide-group__wrapper',
        directives: [{
          name: 'touch',
          value: {
            start: (e: TouchEvent) => this.overflowCheck(e, this.onTouchStart),
            move: (e: TouchEvent) => this.overflowCheck(e, this.onTouchMove),
            end: (e: TouchEvent) => this.overflowCheck(e, this.onTouchEnd)
          }
        }],
        ref: 'wrapper'
      }, [this.genContent()])
    },
    newOffset /* istanbul ignore next */ (direction: 'prepend' | 'append') {
      // Force reflow
      const clientWidth = this.$refs.wrapper.clientWidth

      if (direction === 'prepend') {
        return Math.max(this.scrollOffset - clientWidth, 0)
      }

      return Math.min(
        this.scrollOffset + clientWidth,
        this.$refs.content.clientWidth - clientWidth
      )
    },
    onAffixClick (location: 'prepend' | 'append') {
      this.$emit(`click:${location}`)
      this.scrollTo(location)
    },
    onResize () {
      /* istanbul ignore next */
      if (this._isDestroyed) return

      this.setWidths()
    },
    onTouchStart (e: TouchEvent) {
      const { content } = this.$refs

      this.startX = this.scrollOffset + e.touchstartX as number

      content.style.setProperty('transition', 'none')
      content.style.setProperty('willChange', 'transform')
    },
    onTouchMove (e: TouchEvent) {
      this.scrollOffset = this.startX - e.touchmoveX
    },
    onTouchEnd () {
      const { content, wrapper } = this.$refs
      const maxScrollOffset = content.clientWidth - wrapper.clientWidth

      content.style.setProperty('transition', null)
      content.style.setProperty('willChange', null)

      /* istanbul ignore else */
      if (this.scrollOffset < 0 || !this.isOverflowing) {
        this.scrollOffset = 0
      } else if (this.scrollOffset >= maxScrollOffset) {
        this.scrollOffset = maxScrollOffset
      }
    },
    overflowCheck (e: TouchEvent, fn: (e: TouchEvent) => void) {
      this.isOverflowing && fn(e)
    },
    scrollTo /* istanbul ignore next */ (location: 'prepend' | 'append') {
      this.scrollOffset = this.newOffset(location)
    },
    setOverflow () {
      this.isOverflowing = this.widths.wrapper < this.widths.content
    },
    setWidths () {
      const { content, wrapper } = this.$refs

      this.widths = {
        content: content ? content.clientWidth : 0,
        wrapper: wrapper ? wrapper.clientWidth : 0
      }
    }
  },

  render (h): VNode {
    return h('div', {
      staticClass: 'v-item-group v-slide-group',
      class: this.classes,
      directives: [{
        name: 'resize',
        value: this.onResize
      }]
    }, [
      this.genPrepend(),
      this.genWrapper(),
      this.genAppend()
    ])
  }
})

export default BaseSlideGroup.extend({
  name: 'v-slide-group',

  provide (): object {
    return {
      slideGroup: this
    }
  }
})
