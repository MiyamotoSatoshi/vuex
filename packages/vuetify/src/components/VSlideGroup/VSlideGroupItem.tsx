// Composables
import { makeGroupItemProps, useGroupItem } from '@/composables/group'
import { VSlideGroupSymbol } from './VSlideGroup'

// Utilities
import { defineComponent } from '@/util'

export const VSlideGroupItem = defineComponent({
  name: 'VSlideGroupItem',

  props: {
    ...makeGroupItemProps(),
  },

  setup (props, { slots }) {
    const slideGroupItem = useGroupItem(props, VSlideGroupSymbol)

    return () => slots.default?.({
      isSelected: slideGroupItem.isSelected.value,
      select: slideGroupItem.select,
      toggle: slideGroupItem.toggle,
      selectedClass: slideGroupItem.selectedClass.value,
    })
  },
})
