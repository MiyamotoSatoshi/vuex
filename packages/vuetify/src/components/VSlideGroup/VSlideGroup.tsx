// Styles
import './VSlideGroup.sass'

// Components
import { VFadeTransition } from '@/components/transitions'
import { VIcon } from '@/components/VIcon'

// Composables
import { IconValue } from '@/composables/icons'
import { makeGroupProps, useGroup } from '@/composables/group'
import { makeTagProps } from '@/composables/tag'
import { useDisplay } from '@/composables'
import { useResizeObserver } from '@/composables/resizeObserver'
import { useRtl } from '@/composables/rtl'

// Utilities
import { bias, calculateCenteredOffset, calculateUpdatedOffset } from './helpers'
import { clamp, defineComponent, IN_BROWSER, useRender } from '@/util'
import { computed, ref, watch } from 'vue'

// Types
import type { GroupProvide } from '@/composables/group'
import type { InjectionKey } from 'vue'

export const VSlideGroupSymbol: InjectionKey<GroupProvide> = Symbol.for('vuetify:v-slide-group')

export const VSlideGroup = defineComponent({
  name: 'VSlideGroup',

  props: {
    centerActive: Boolean,
    direction: {
      type: String,
      default: 'horizontal',
    },
    symbol: {
      type: null,
      default: VSlideGroupSymbol,
    },
    nextIcon: {
      type: IconValue,
      default: '$next',
    },
    prevIcon: {
      type: IconValue,
      default: '$prev',
    },
    showArrows: {
      type: [Boolean, String],
      validator: (v: any) => (
        typeof v === 'boolean' || [
          'always',
          'desktop',
          'mobile',
        ].includes(v)
      ),
    },

    ...makeTagProps(),
    ...makeGroupProps({
      selectedClass: 'v-slide-group-item--active',
    }),
  },

  emits: {
    'update:modelValue': (value: any) => true,
  },

  setup (props, { slots }) {
    const { isRtl } = useRtl()
    const { mobile } = useDisplay()
    const group = useGroup(props, props.symbol)
    const isOverflowing = ref(false)
    const scrollOffset = ref(0)
    const containerSize = ref(0)
    const contentSize = ref(0)
    const isHorizontal = computed(() => props.direction === 'horizontal')

    const { resizeRef: containerRef, contentRect: containerRect } = useResizeObserver()
    const { resizeRef: contentRef, contentRect } = useResizeObserver()

    const firstSelectedIndex = computed(() => {
      if (!group.selected.value.length) return -1

      return group.items.value.findIndex(item => item.id === group.selected.value[0])
    })

    const lastSelectedIndex = computed(() => {
      if (!group.selected.value.length) return -1

      return group.items.value.findIndex(item => item.id === group.selected.value[group.selected.value.length - 1])
    })

    if (IN_BROWSER) {
      let frame = -1
      watch(() => [group.selected.value, containerRect.value, contentRect.value, isHorizontal.value], () => {
        cancelAnimationFrame(frame)
        frame = requestAnimationFrame(() => {
          if (containerRect.value && contentRect.value) {
            const sizeProperty = isHorizontal.value ? 'width' : 'height'

            containerSize.value = containerRect.value[sizeProperty]
            contentSize.value = contentRect.value[sizeProperty]

            isOverflowing.value = containerSize.value + 1 < contentSize.value
          }

          if (firstSelectedIndex.value >= 0 && contentRef.value) {
            // TODO: Is this too naive? Should we store element references in group composable?
            const selectedElement = contentRef.value.children[lastSelectedIndex.value] as HTMLElement

            if (firstSelectedIndex.value === 0 || !isOverflowing.value) {
              scrollOffset.value = 0
            } else if (props.centerActive) {
              scrollOffset.value = calculateCenteredOffset({
                selectedElement,
                containerSize: containerSize.value,
                contentSize: contentSize.value,
                isRtl: isRtl.value,
                isHorizontal: isHorizontal.value,
              })
            } else if (isOverflowing.value) {
              scrollOffset.value = calculateUpdatedOffset({
                selectedElement,
                containerSize: containerSize.value,
                contentSize: contentSize.value,
                isRtl: isRtl.value,
                currentScrollOffset: scrollOffset.value,
                isHorizontal: isHorizontal.value,
              })
            }
          }
        })
      })
    }

    const disableTransition = ref(false)

    let startTouch = 0
    let startOffset = 0

    function onTouchstart (e: TouchEvent) {
      const sizeProperty = isHorizontal.value ? 'clientX' : 'clientY'
      const sign = isRtl.value && isHorizontal.value ? -1 : 1
      startOffset = sign * scrollOffset.value
      startTouch = e.touches[0][sizeProperty]
      disableTransition.value = true
    }

    function onTouchmove (e: TouchEvent) {
      if (!isOverflowing.value) return

      const sizeProperty = isHorizontal.value ? 'clientX' : 'clientY'
      const sign = isRtl.value && isHorizontal.value ? -1 : 1
      scrollOffset.value = sign * (startOffset + startTouch - e.touches[0][sizeProperty])
    }

    function onTouchend (e: TouchEvent) {
      const maxScrollOffset = contentSize.value - containerSize.value

      if (scrollOffset.value < 0 || !isOverflowing.value) {
        scrollOffset.value = 0
      } else if (scrollOffset.value >= maxScrollOffset) {
        scrollOffset.value = maxScrollOffset
      }

      disableTransition.value = false
    }

    function onScroll () {
      if (!containerRef.value) return

      containerRef.value[isHorizontal.value ? 'scrollLeft' : 'scrollTop'] = 0
    }

    const isFocused = ref(false)
    function onFocusin (e: FocusEvent) {
      isFocused.value = true

      if (!isOverflowing.value || !contentRef.value) return

      // Focused element is likely to be the root of an item, so a
      // breadth-first search will probably find it in the first iteration
      for (const el of e.composedPath()) {
        for (const item of contentRef.value.children) {
          if (item === el) {
            scrollOffset.value = calculateUpdatedOffset({
              selectedElement: item as HTMLElement,
              containerSize: containerSize.value,
              contentSize: contentSize.value,
              isRtl: isRtl.value,
              currentScrollOffset: scrollOffset.value,
              isHorizontal: isHorizontal.value,
            })
            return
          }
        }
      }
    }

    function onFocusout (e: FocusEvent) {
      isFocused.value = false
    }

    function onFocus (e: FocusEvent) {
      if (
        !isFocused.value &&
        !(e.relatedTarget && contentRef.value?.contains(e.relatedTarget as Node))
      ) focus()
    }

    function onKeydown (e: KeyboardEvent) {
      if (!contentRef.value) return

      if (isHorizontal.value) {
        if (e.key === 'ArrowRight') {
          focus(isRtl.value ? 'prev' : 'next')
        } else if (e.key === 'ArrowLeft') {
          focus(isRtl.value ? 'next' : 'prev')
        }
      } else {
        if (e.key === 'ArrowDown') {
          focus('next')
        } else if (e.key === 'ArrowUp') {
          focus('prev')
        }
      }

      if (e.key === 'Home') {
        focus('first')
      } else if (e.key === 'End') {
        focus('last')
      }
    }

    function focus (location?: 'next' | 'prev' | 'first' | 'last') {
      if (!contentRef.value) return

      if (!location) {
        contentRef.value.querySelector('[tabindex]')
        const focusable = [...contentRef.value.querySelectorAll(
          'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'
        )].filter(el => !el.hasAttribute('disabled')) as HTMLElement[]
        focusable[0]?.focus()
      } else if (location === 'next') {
        const el = contentRef.value.querySelector(':focus')?.nextElementSibling as HTMLElement | undefined
        if (el) el.focus()
        else focus('first')
      } else if (location === 'prev') {
        const el = contentRef.value.querySelector(':focus')?.previousElementSibling as HTMLElement | undefined
        if (el) el.focus()
        else focus('last')
      } else if (location === 'first') {
        (contentRef.value.firstElementChild as HTMLElement)?.focus()
      } else if (location === 'last') {
        (contentRef.value.lastElementChild as HTMLElement)?.focus()
      }
    }

    function scrollTo (location: 'prev' | 'next') {
      const newAbsoluteOffset = scrollOffset.value + (location === 'prev' ? -1 : 1) * containerSize.value

      scrollOffset.value = clamp(newAbsoluteOffset, 0, contentSize.value - containerSize.value)
    }

    const contentStyles = computed(() => {
      // This adds friction when scrolling the 'wrong' way when at max offset
      let scrollAmount = scrollOffset.value > contentSize.value - containerSize.value
        ? -(contentSize.value - containerSize.value) + bias(contentSize.value - containerSize.value - scrollOffset.value)
        : -scrollOffset.value

      // This adds friction when scrolling the 'wrong' way when at min offset
      if (scrollOffset.value <= 0) {
        scrollAmount = bias(-scrollOffset.value)
      }

      const sign = isRtl.value && isHorizontal.value ? -1 : 1
      return {
        transform: `translate${isHorizontal.value ? 'X' : 'Y'}(${sign * scrollAmount}px)`,
        transition: disableTransition.value ? 'none' : '',
        willChange: disableTransition.value ? 'transform' : '',
      }
    })

    const slotProps = computed(() => ({
      next: group.next,
      prev: group.prev,
      select: group.select,
      isSelected: group.isSelected,
    }))

    const hasAffixes = computed(() => {
      switch (props.showArrows) {
        // Always show arrows on desktop & mobile
        case 'always': return true

        // Always show arrows on desktop
        case 'desktop': return !mobile.value

        // Show arrows on mobile when overflowing.
        // This matches the default 2.2 behavior
        case true: return isOverflowing.value || Math.abs(scrollOffset.value) > 0

        // Always show on mobile
        case 'mobile': return (
          mobile.value ||
          (isOverflowing.value || Math.abs(scrollOffset.value) > 0)
        )

        // https://material.io/components/tabs#scrollable-tabs
        // Always show arrows when
        // overflowed on desktop
        default: return (
          !mobile.value &&
          (isOverflowing.value || Math.abs(scrollOffset.value) > 0)
        )
      }
    })

    const hasPrev = computed(() => {
      return Math.abs(scrollOffset.value) > 0
    })

    const hasNext = computed(() => {
      // Check one scroll ahead to know the width of right-most item
      return contentSize.value > Math.abs(scrollOffset.value) + containerSize.value
    })

    useRender(() => (
      <props.tag
        class={[
          'v-slide-group',
          {
            'v-slide-group--vertical': !isHorizontal.value,
            'v-slide-group--has-affixes': hasAffixes.value,
            'v-slide-group--is-overflowing': isOverflowing.value,
          },
        ]}
        tabindex={ (isFocused.value || group.selected.value.length) ? -1 : 0 }
        onFocus={ onFocus }
      >
        { hasAffixes.value && (
          <div
            key="prev"
            class={[
              'v-slide-group__prev',
              { 'v-slide-group__prev--disabled': !hasPrev.value },
            ]}
            onClick={ () => scrollTo('prev') }
          >
            { slots.prev?.(slotProps.value) ?? (
              <VFadeTransition>
                <VIcon icon={ isRtl.value ? props.nextIcon : props.prevIcon }></VIcon>
              </VFadeTransition>
            ) }
          </div>
        ) }

        <div
          key="container"
          ref={ containerRef }
          class="v-slide-group__container"
          onScroll={ onScroll }
        >
          <div
            ref={ contentRef }
            class="v-slide-group__content"
            style={ contentStyles.value }
            onTouchstartPassive={ onTouchstart }
            onTouchmovePassive={ onTouchmove }
            onTouchendPassive={ onTouchend }
            onFocusin={ onFocusin }
            onFocusout={ onFocusout }
            onKeydown={ onKeydown }
          >
            { slots.default?.(slotProps.value) }
          </div>
        </div>

        { hasAffixes.value && (
          <div
            key="next"
            class={[
              'v-slide-group__next',
              { 'v-slide-group__next--disabled': !hasNext.value },
            ]}
            onClick={ () => scrollTo('next') }
          >
            { slots.next?.(slotProps.value) ?? (
              <VFadeTransition>
                <VIcon icon={ isRtl.value ? props.prevIcon : props.nextIcon }></VIcon>
              </VFadeTransition>
            ) }
          </div>
        ) }
      </props.tag>
    ))

    return {
      selected: group.selected,
      scrollTo,
      scrollOffset,
      focus,
    }
  },
})

export type VSlideGroup = InstanceType<typeof VSlideGroup>
