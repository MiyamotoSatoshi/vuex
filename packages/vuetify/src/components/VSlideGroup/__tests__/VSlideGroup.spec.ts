// Components
import VSlideGroup from '../VSlideGroup'

// Utilities
import { ExtractVue } from '../../../util/mixins'
import {
  mount,
  shallowMount,
  Wrapper
} from '@vue/test-utils'

describe('VSlideGroup.ts', () => {
  type Instance = ExtractVue<typeof VSlideGroup>
  let mountFunction: (options?: object) => Wrapper<Instance>

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return shallowMount(VSlideGroup, {
        methods: { setWidths: jest.fn() },
        mocks: {
          $vuetify: {
            rtl: false,
            breakpoint: {
              width: 1920
            }
          }
        },
        ...options
      })
    }
  })

  it('should conditionally have affixes, prev and next', () => {
    const wrapper = mountFunction({
      data: () => ({
        isOverflowing: true
      }),
      propsData: {
        showArrows: true
      }
    })

    expect(wrapper.vm.hasAffixes).toBe(true)
    expect(wrapper.vm.hasNext).toBe(false)
    expect(wrapper.vm.hasPrev).toBe(false)

    wrapper.setData({
      scrollOffset: 100,
      widths: {
        content: 1000,
        wrapper: 500
      }
    })

    expect(wrapper.vm.hasPrev).toBe(true)

    wrapper.setData({
      scrollOffset: -100,
      widths: {
        content: 1000,
        wrapper: 500
      }
    })

    expect(wrapper.vm.hasNext).toBe(true)
  })

  it('should be considered mobile', async () => {
    const wrapper = mountFunction()

    expect(wrapper.vm.isMobile).toBe(false)

    wrapper.vm.$vuetify.breakpoint.width = 700

    expect(wrapper.vm.isMobile).toBe(true)
  })

  // TODO: Unsure what we're actually testing, willChange not found in jest 24
  it.skip('should call on touch methods', async () => {
    const wrapper = mountFunction({
      data: () => ({
        isOverflowing: true
      })
    })

    expect(wrapper.vm.scrollOffset).toBe(0)

    const touchstartEvent = {
      touchstartX: 10,
      touchmoveX: 0
    }

    wrapper.vm.onTouchStart(touchstartEvent)

    expect(wrapper.vm.startX).toBe(10)
    expect(wrapper.vm.$refs.content.style.transition).toBe('none')
    expect(wrapper.vm.$refs.content.style.willChange).toBe('transform')

    const touchmoveEvent = {
      touchstartX: 10,
      touchmoveX: 100
    }
    wrapper.vm.onTouchMove(touchmoveEvent)

    expect(wrapper.vm.scrollOffset).toBe(-90)

    wrapper.vm.onTouchEnd()

    expect(wrapper.vm.scrollOffset).toBe(0)

    wrapper.setData({
      scrollOffset: 90,
      isOverflowing: true
    })

    wrapper.vm.onTouchEnd()
    expect(wrapper.vm.scrollOffset).toBe(0)

    // TODO: Figure out why this doesn't work in TS + vue-test-utils
    // groupWrapper.trigger('touchmove')
    // touch(groupWrapper).start(0, 0)
    // touch(groupWrapper).end(0, 0)
    // touch(groupWrapper).move(15, 15)
    // expect(onTouch.mock.calls.length).toBe(3)
  })

  it('should invoke method only if overflowing', () => {
    const wrapper = mountFunction()
    const fn = jest.fn()
    const event = {
      touchstartX: 0,
      touchmoveX: 0
    }

    wrapper.vm.overflowCheck(event, fn)
    expect(fn).not.toHaveBeenCalled()

    wrapper.setData({ isOverflowing: true })
    wrapper.vm.overflowCheck(event, fn)
    expect(fn).toHaveBeenCalled()
  })

  it('it should scroll from affix click', async () => {
    const onClick = jest.fn()
    const scrollTo = jest.fn()
    const setWidths = jest.fn()
    const wrapper = mountFunction({
      methods: { scrollTo, setWidths },
      propsData: {
        showArrows: true
      },
      listeners: {
        'click:prev': onClick,
        'click:next': onClick
      }
    })

    wrapper.setData({
      isOverflowing: true,
      scrollOffset: 200,
      widths: {
        content: 1000,
        wrapper: 500
      }
    })

    await wrapper.vm.$nextTick()

    const prev = wrapper.find('.v-slide-group__prev')
    const next = wrapper.find('.v-slide-group__next')

    prev.trigger('click')
    next.trigger('click')
    expect(scrollTo).toHaveBeenCalledTimes(2)
    expect(onClick).toHaveBeenCalledTimes(2)
  })

  it('should accept scoped slots', () => {
    const wrapper = mount(VSlideGroup, {
      computed: {
        hasAffixes: () => true,
        hasNext: () => true,
        hasPrev: () => true
      },
      propsData: {
        showArrows: true
      },
      scopedSlots: {
        prev () {
          return this.$createElement('div', {
            staticClass: 'fizz'
          }, 'foo')
        },
        next () {
          return this.$createElement('div', {
            staticClass: 'fizz'
          }, 'bar')
        }
      }
    })

    wrapper.setData({ isOverflowing: true })

    expect(wrapper.findAll('.fizz')).toHaveLength(2)
  })

  it('should match snapshot in rtl', async () => {
    const wrapper = mountFunction({
      computed: {
        hasAffixes: () => true,
        hasNext: () => true,
        hasPrev: () => true
      },
      propsData: {
        showArrows: true
      },
      mocks: {
        $vuetify: {
          rtl: true
        }
      }
    })

    const html1 = wrapper.html()

    expect(html1).toMatchSnapshot()

    wrapper.vm.$vuetify.rtl = false

    const html2 = wrapper.html()

    expect(html1).not.toEqual(html2)
    expect(html2).toMatchSnapshot()
  })
})
