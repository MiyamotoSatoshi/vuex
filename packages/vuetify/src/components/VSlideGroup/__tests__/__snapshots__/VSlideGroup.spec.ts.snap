// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VSliderGroup.ts should match snapshot in rtl 1`] = `

<div class="v-item-group v-slide-group theme--light">
  <div class="v-slide-group__prev">
    <fade-transition-stub origin="top center 0">
      <vuecomponent-stub>
        $vuetify.icons.next
      </vuecomponent-stub>
    </fade-transition-stub>
  </div>
  <div class="v-slide-group__wrapper">
    <div class="v-slide-group__content">
    </div>
  </div>
  <div class="v-slide-group__next">
    <fade-transition-stub origin="top center 0">
      <vuecomponent-stub>
        $vuetify.icons.prev
      </vuecomponent-stub>
    </fade-transition-stub>
  </div>
</div>

`;

exports[`VSliderGroup.ts should match snapshot in rtl 2`] = `

<div class="v-item-group v-slide-group theme--light">
  <div class="v-slide-group__prev">
    <fade-transition-stub origin="top center 0">
      <vuecomponent-stub>
        $vuetify.icons.prev
      </vuecomponent-stub>
    </fade-transition-stub>
  </div>
  <div class="v-slide-group__wrapper">
    <div class="v-slide-group__content">
    </div>
  </div>
  <div class="v-slide-group__next">
    <fade-transition-stub origin="top center 0">
      <vuecomponent-stub>
        $vuetify.icons.next
      </vuecomponent-stub>
    </fade-transition-stub>
  </div>
</div>

`;
